src/unix/macos.rs

```diff
@@ -0,0 +1,117 @@
+//! Beginning with MacOS 10.3, `utimensat` is supported by the MacOS, so here, we check if the symbol exists
+//! and if not, we fallabck to `utimes`.
+use crate::FileTime;
+use std::ffi::CString;
+use std::fs::{File, Metadata};
+use std::os::unix::prelude::*;
+use std::path::Path;
+use std::{io, ptr};
+
+pub fn set_file_times(p: &Path, atime: FileTime, mtime: FileTime) -> io::Result<()> {
+    set_times(p, Some(atime), Some(mtime), false)
+}
+
+pub fn set_file_mtime(p: &Path, mtime: FileTime) -> io::Result<()> {
+    set_times(p, None, Some(mtime), false)
+}
+
+pub fn set_file_atime(p: &Path, atime: FileTime) -> io::Result<()> {
+    set_times(p, Some(atime), None, false)
+}
+
+fn utimensat_available() -> io::Result<bool> {
+    let lib_name = CString::new("libSystem.dylib".as_bytes())?;
+    let handle = unsafe { libc::dlopen(lib_name.as_ptr() as *const _, 0) };
+    if handle == ptr::null_mut() {
+        return Err(io::Error::last_os_error());
+    }
+    let symbol_name = CString::new("utimensat".as_bytes())?;
+    let address = unsafe { libc::dlsym(handle, symbol_name.as_ptr() as *const _) };
+    if address == ptr::null_mut() {
+        Ok(false)
+    } else {
+        Ok(true)
+    }
+}
+
+pub fn set_file_handle_times(
+    f: &File,
+    atime: Option<FileTime>,
+    mtime: Option<FileTime>,
+) -> io::Result<()> {
+    if utimensat_available()? {
+        let (atime, mtime) = match get_times(atime, mtime, || f.metadata())? {
+            Some(pair) => pair,
+            None => return Ok(()),
+        };
+        let times = [
+            super::to_timespec(&Some(atime)),
+            super::to_timespec(&Some(mtime)),
+        ];
+        let rc = unsafe { libc::futimens(f.as_raw_fd(), times.as_ptr()) };
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(io::Error::last_os_error())
+        }
+    } else {
+        super::utimes::set_file_handle_times(f, atime, mtime)
+    }
+}
+
+fn get_times(
+    atime: Option<FileTime>,
+    mtime: Option<FileTime>,
+    current: impl FnOnce() -> io::Result<Metadata>,
+) -> io::Result<Option<(FileTime, FileTime)>> {
+    let pair = match (atime, mtime) {
+        (Some(a), Some(b)) => (a, b),
+        (None, None) => return Ok(None),
+        (Some(a), None) => {
+            let meta = current()?;
+            (a, FileTime::from_last_modification_time(&meta))
+        }
+        (None, Some(b)) => {
+            let meta = current()?;
+            (FileTime::from_last_access_time(&meta), b)
+        }
+    };
+    Ok(Some(pair))
+}
+
+pub fn set_symlink_file_times(p: &Path, atime: FileTime, mtime: FileTime) -> io::Result<()> {
+    set_times(p, Some(atime), Some(mtime), true)
+}
+
+fn set_times(
+    p: &Path,
+    atime: Option<FileTime>,
+    mtime: Option<FileTime>,
+    symlink: bool,
+) -> io::Result<()> {
+    if utimensat_available()? {
```

Comment:
This boolean evaluation can actually be pretty expensive, so we'll want to cache the result of `dlsym` into a static (probably some sort of atomic) for making further calls to this function quick